### ISSUE TITLE
Alternative database format?

### ISSUE THREAD

**Issue by judgej** [gaubert/gmvault#204]
Would (or could) any alternative database formats be considered or possible? Backing up an account with many 10s of thousands of emails quickly eats into disk space on many file systems. Each email results in two files, often both 1kbyte or less. A storage format that combines files, I feel, would make storage a lot more efficient.

For example, I have one account that backs up to a 10Gbyte database using MailStore on Windows. Based on the FS block size, that would need more like 30Gbyte using gmvault.

Compressing the files does not actually help with storage space used, when the files are already so small. But combining them into multi-file archives certainly would.

Is this something that has ever been considered?

**Comment by judgej** [comment]
I know little about the architecture of this application is at this stage, so have no idea if storage is a simple layer that can be switched in and out with alternatives in an extendable way.

**Comment by gaubert** [comment]
@judgej You are right and it is something to consider. In general Gmail backup stay pretty small few GB which nowadays are easy to handle. I am considering changing anyway the archive format for V2.0

**Comment by judgej** [comment]
Cool. I know little python, so cannot help much during development. However, I am here for testing.

**Comment by jvhaarst** [comment]
I just did a search for options:
* [mbox](https://en.wikipedia.org/wiki/Mbox) : Single file, reclaiming space from deleted messages is expensive.
* [Maildir](https://en.wikipedia.org/wiki/Maildir),[MIX](https://en.wikipedia.org/wiki/MIX_(Email)),[mh](https://en.wikipedia.org/wiki/MH_Message_Handling_System), : Multiple files, so no real change with current situation.
* [sqlite](https://www.sqlite.org), as used by e.g. Apple Mail ([schema](http://labs.wordtothewise.com/mailapp/mailapp/))

Thunderbird also wanted to be able to store mail in different formats, so they implemented [Pluggable Mail stores](https://wiki.mozilla.org/Thunderbird:Pluggable_Mail_Stores), but I  couldn't find any implementations beyond mbox and Maildir.

My vote would be for [sqlite3](http://www.sqlite.org/appfileformat.html), but with a simpler schema than Apple Mail as gmvault only needs to store data, and isn't interested in the links between the mails, right ? 
[Full text search](https://www.sqlite.org/fts3.html) is something that might be interesting, although that would increase the size of the database.

**Comment by NewsGuyTor** [comment]
Another big advantage of a new database format would be to have way less small files, which massively decrease syncing/transfer speeds.

**Comment by jvhaarst** [comment]
@csirmaz has already started work on this, look e.g. here :
 https://github.com/csirmaz/gmvault/commit/8cc70fccc36b1dcf60f857191d7fa7c55c3b853d

He doesn't store the mails itself in the database yet.

**Comment by Alex131089** [comment]
Some time ago I had the idea for a similar idea too: instead of having a huge number of .gz files, an unique / a few (per month?) archive files containing the .eml/the account could be more efficient.
At first I started looking for apps allowing to mount an archive, but is seems there's no acceptable apps.
So another way could be the use of a python lib with an efficient archive format like 7z, so gmvault would create/open the db/archive file, and do it's works directly in it, instead of the FS.
But it seems the main libs (https://pypi.org/search/?q=libarchive) are not directly usables:
* https://pypi.org/project/libarchive/ doesn't allow to add files from memory (yet) (and I don't think using temporary files would be efficient)
* https://pypi.org/project/libarchive-c/ needs some work to work on Windows

If that can give more ideas / leads. And it might be simpler than a full new format.
